presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SingleDistinctOptimizer.java

```diff
@@ -0,0 +1,142 @@
+            if (masks.size() != 1 || node.getMasks().size() != node.getAggregations().size()) {
+                return context.defaultRewrite(node, Optional.empty());
+            }
+            Symbol mask = Iterables.getOnlyElement(masks);
+            PlanNode source = context.rewrite(node.getSource(), Optional.of(mask));
+
+            Map<Symbol, FunctionCall> aggregations = ImmutableMap.copyOf(Maps.transformValues(node.getAggregations(), call -> new FunctionCall(call.getName(), call.getWindow(), false, call.getArguments())));
+
+            return new AggregationNode(idAllocator.getNextId(),
+                                        source,
+                                        node.getGroupBy(),
+                                        aggregations,
+                                        node.getFunctions(),
+                                        Collections.emptyMap(),
+                                        node.getStep(),
+                                        node.getSampleWeight(),
+                                        node.getConfidence(),
+                                        node.getHashSymbol());
+        }
+
+        public PlanNode visitMarkDistinct(MarkDistinctNode node, RewriteContext<Optional<Symbol>> context)
+        {
+            Optional<Symbol> mask = context.get();
+            if (mask.isPresent() && mask.get().equals(node.getMarkerSymbol())) {
+                // rewrite Distinct into GroupBy
+                AggregationNode aggregationNode = new AggregationNode(idAllocator.getNextId(),
+                                                                        node.getSource(),
+                                                                        node.getDistinctSymbols(),
+                                                                        Collections.emptyMap(),
+                                                                        Collections.emptyMap(),
+                                                                        Collections.emptyMap(),
+                                                                        SINGLE,
+                                                                        Optional.empty(),
+                                                                        1.0,
+                                                                        node.getHashSymbol());
+
+                ImmutableMap.Builder<Symbol, Expression> outputSymbols = ImmutableMap.builder();
+                for (Symbol symbol : node.getDistinctSymbols()) {
+                    Expression expression = new QualifiedNameReference(symbol.toQualifiedName());
+                    outputSymbols.put(symbol, expression);
+                }
+
+                if (node.getHashSymbol().isPresent()) {
+                    Symbol hashSymbol = node.getHashSymbol().get();
+                    Expression hashExpression = getHashExpression(ImmutableList.of(hashSymbol));
```

Comment:
Did you see my previous comment about this? I don't think you need this, just project the hashSymbol as is.